### ISSUE TITLE
State easing

### ISSUE THREAD

**Issue by scottbedard** [antfu/vueuse#62]
While I'm in here, would you mind if I added one more util to the library? This one would be for transitioning state along [cubic bezier curves](https://easings.net/). Below is an example of what I'm proposing...

```vue
<template>
  <div>
    <input v-model.number="baseNumber" /> {{ number }}
  </div>
</template>

<script>
export default {
  setup() {
    const baseNumber = ref(0)

    const number = useStateTransition(baseNumber, {
      duration: 1000,
      steps: 30, // probably a reasonable default, but allows for custom framerates
      transition: [0.77, 0, 0.175, 1] // maybe have aliases like 'easeInOutQuart'?
    })

    return { baseNumber, number }
  }
}
</script>
```

**Comment by scottbedard** [comment]
And just for fun, an example of using the util to transition colors, and maybe some shorthand?

```vue
<template>
  <div>
    <input v-model.number="red" type="range" min="0" max="255" />
    <input v-model.number="green" type="range" min="0" max="255" />
    <input v-model.number="blue" type="range" min="0" max="255" />

    <h1 :style="{ color }">Hello world</h1>
  </div>
</template>

<script>
export default {
  setup() {
    const baseRed = ref(0)
    const baseGreen = ref(0)
    const baseBlue = ref(0)

    const [red, green, blue] = useStateTransition([baseRed, baseGreen, baseBlue], {
      duration: 1000,
      transition: 'easeInOutCubic'
    })

    const color = computed(() => `#${red}${green}${blue}`)

    return {
      baseRed,
      baseGreen,
      baseBlue,
      color
    }
  }
}
</script>
```

**Comment by antfu** [comment]
This looks very useful, but my concern is that this can be over complicated to implement well and may surpass the goal of VueUse. Libraries like [tween.js](https://github.com/tweenjs/tween.js/) are much more mature and flexible to this, maybe we could add a wrapper addon for that?

**Comment by scottbedard** [comment]
I don't think it's necessary to pull in additional libraries. I've thrown together a proof of concept that I can push up in just a few minutes, just cleaning some things up right now. If you look at it and decide you'd rather use a library like Tween, no worries, just go ahead and close the PR when I open it 🙂

**Comment by scottbedard** [comment]
@antfu ok I've pushed up what I've done. [This is the only complicated bit](https://github.com/antfu/vueuse/pull/63/files#diff-568e52af26345056c37625165c10836aR15-R38), it creates a function that we can use to get the corresponding value at various points in the transition. Let me know if you're ok with this amount of complexity in this library, because if so there are a couple changes I want to make before merging.

It's mostly just cleanup, but I'd also like to refactor it to use an raf loop instead of timeouts for smoother transitioning and getting rid of the "steps" concept.

**Comment by scottbedard** [comment]
@antfu ok, we should be good to go now. I've made the refactor to a raf loop. I'm pleased with how it came out, the transition is much smoother now. 

I made a couple other changes, let me know if you have any opinions on these...
1. Renamed "easing" to "transition", I think it's a bit more descriptive of what this utility is for
2. If the base number changes while state is still transitioning, the new transition will just start from wherever the previous one left off
3. It no longer rounds transitioned value. It occurred to me that people will likely want to use this with monetary values, so we should let them decide how to round things

Let me know if there is anything else you need from me 🙂

**Comment by antfu** [comment]
Shipped in v0.0.23. Thanks :)

**Comment by scottbedard** [comment]
Awesome, thanks for letting me wedge those two utils into this library 😄
